pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,231 @@
+				log.Info("queueing ingress canary", "related", a.Meta.GetSelfLink())
+				return []reconcile.Request{
+					{
+						NamespacedName: types.NamespacedName{
+							Namespace: namespace,
+							Name:      a.Meta.GetName(),
+						},
+					},
+				}
+			} else {
+				return []reconcile.Request{}
+			}
+		}),
+	}
+}
+
+// enqueueRequestForDefaultIngressController returns canary controller
+// reconcile requests for the default ingress controller.
+func enqueueRequestForDefaultIngressController(namespace string) handler.EventHandler {
+	return &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+			if cmp.Equal(a.Meta.GetName(), manifests.DefaultIngressControllerName) {
+				log.Info("queueing ingress canary", "related", a.Meta.GetSelfLink())
+				return []reconcile.Request{
+					{
+						NamespacedName: types.NamespacedName{
+							Namespace: namespace,
+							Name:      manifests.DefaultIngressControllerName,
+						},
+					},
+				}
+			} else {
+				return []reconcile.Request{}
+			}
+		}),
+	}
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	Namespace   string
+	CanaryImage string
+}
+
+// reconciler handles the actual canary reconciliation logic in response to
+// events.
+type reconciler struct {
+	Config
+
+	client client.Client
+	cache  cache.Cache
+}
+
+// Switch the current RoutePort that the route points to.
+// Use this function to periodically update the canary route endpoint
+// to verify if the router has wedged.
+func (r *reconciler) rotateRouteEndpoint(service *corev1.Service, current *routev1.Route) (*routev1.Route, error) {
+
+	updated, err := chooseRandomServicePort(service, current)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to rotate route port: %v", err)
+	}
+
+	// update route resource here
+	_, err = r.updateCanaryRoute(current, updated)
+	if err != nil {
+		return current, err
+	}
+
+	return updated, nil
+}
+
+// chooseRandomService returns a route resource with Spec.Port set to a random
+// port selected from service.Spec.Ports, exlcuding route.Spec.Port to gaurantee
+// that a new port is selected.
+func chooseRandomServicePort(service *corev1.Service, route *routev1.Route) (*routev1.Route, error) {
+	servicePorts := service.Spec.Ports
+	currentPort := route.Spec.Port
+
+	numServicePorts := len(servicePorts)
+	if numServicePorts == 0 {
+		return nil, fmt.Errorf("Service has no ports")
+	} else if numServicePorts == 1 {
+		return nil, fmt.Errorf("Service has only one port, no change possible")
+	}
+
+	updated := route.DeepCopy()
+
+	availablePorts := []intstr.IntOrString{}
+	// Filter currentPort from servicePorts
+	for _, port := range servicePorts {
+		if !cmp.Equal(port.TargetPort, currentPort.TargetPort) {
+			availablePorts = append(availablePorts, port.TargetPort)
+		}
+	}
+
+	updated.Spec.Port = &routev1.RoutePort{
+		TargetPort: availablePorts[rand.Intn(len(availablePorts))],
```

Comment:
My thinking was we might end up expanding a canary service to have > 2 ports. But since that isn't going to happen for now, ill convert `chooseRandomServicePort` to `cycleServicePort` as you suggested.